File: javascript/2.add-two-numbers.js
/*
 * @lc app=leetcode id=2 lang=javascript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  let head = currNode = new ListNode()
  let carry = 0
  while (l1 || l2) {
    if (!l1) {
      let sum = l2.val + carry
      currNode.next = new ListNode(sum % 10)
      currNode = currNode.next
      l2 = l2.next
      carry = Math.floor(sum / 10)
    } else if (!l2) {
      let sum = l1.val + carry
      currNode.next = new ListNode(sum % 10)
      currNode = currNode.next
      l1 = l1.next
      carry = Math.floor(sum / 10)
    } else {
      let sum = l1.val + l2.val + carry
      currNode.next = new ListNode(sum % 10)
      currNode = currNode.next
      l1 = l1 ? l1.next : null
      l2 = l2 ? l2.next : null
      carry = Math.floor(sum / 10)
    }
  }
  if (carry > 0) {
    currNode.next = new ListNode(carry)
  }
  return head.next
};
// @lc code=end

File: cpp/2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        // iterate with two pointers
        // add and store carry
        // O(n) time O(k) space
        int sum;
        int carry = 0;
        ListNode *head = new ListNode();
        ListNode *currNode = new ListNode();
        head->next = currNode;
        while (l1 || l2 || carry)
        {
            int firstVal = 0;
            int secondVal = 0;
            if (l1)
            {
                firstVal += l1->val;
                l1 = l1->next;
            }
            if (l2)
            {
                secondVal += l2->val;
                l2 = l2->next;
            }
            sum = firstVal + secondVal + carry;
            ListNode *newNode = new ListNode(sum % 10);
            carry = sum / 10;
            currNode->next = newNode;
            currNode = newNode;
        }
        return head->next->next;
    }
};
// @lc code=end
